### PR TITLE
Fix filtering by unbounded numeric array literal, `= <numeric_array>::numeric[]`

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Follows https://github.com/crate/crate/pull/16862.

Fixes:
```
cr> create table t (a numeric(4,2)[]);
CREATE OK, 1 row affected (1.651 sec)
cr> insert into t values ([1.11]);
INSERT OK, 1 row affected (0.045 sec)
cr> select * from t where a = [1.111]::numeric[];
+--------+
| a      |
+--------+
| [1.11] |
+--------+
SELECT 1 row in set (0.144 sec)
```

https://github.com/crate/crate/blob/a1f77cfb706fb3fb470a58e5d0dad58eb679ade6/docs/appendices/release-notes/5.9.1.rst?plain=1#L118-L120 is enough that we do not need an extra release entry.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/lucene/NumericEqQueryTest.java [hunk 1]
```diff
@@ -144,4 +154,19 @@ public void test_uses_binary_encoded_range_queries_for_large_numeric() throws Ex
         assertThat(query).isInstanceOf(PointRangeQuery.class);
         assertThat(query.toString()).isEqualTo("y:[-99999999999999999999999999999999999999 TO 274679983711617676]");
     }
+
+    @Test
+    public void test_out_of_bounds_numeric_values_are_filtered_from_numeric_array_literal() {
+        // removing 1.111 because it will never match numeric values with scale = 2; 1.110 is kept since it is equivalent to 1.11
+        assertThat(convert("xarr = [1, 1.1, 1.110, 1.111]::numeric[]").toString())
+            .isEqualTo("+xarr:{100 110 111} +(xarr = [1.0, 1.1, 1.11, 1.111])"); // generic query still contains 1.111
```
Comment (jeeminso) [inline server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: This assertion fails with actual return, `+xarr:{100 110 111} +(xarr = [1.00, 1.10, 1.11, 1.11])`. The generic function query contains numeric values look to be casted.

My initial attempt to address this was to add a hacky workaround: https://github.com/crate/crate/pull/16829/files#diff-2eaa286cd80d776d19107cfeb1f4d4ed619d12cb7697c5d4109679888526ca59R77-R80.

Could you share your opinions? @seut @matriv

Comment (jeeminso) [reply server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: Forgot to share, what happens is that as we swap the cast with `SwapCastsInComparisonOperators`, we swap ref.valueType()(numeric(18,2)) not castFunction.valueType()(numeric(null,null)) causing the numeric literal to be casted to numeric(18,2) losing the original scales.

Comment (matriv) [reply server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: Shouldn't we remove the value completely from the generic query?

Comment (matriv) [reply server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: I cannot think of another solution for this one, and couldn't find a case that breaks with this fix in `SwapCastsInComparisonOperators`. Maybe you could also test other variations with `= ANY`, `LIKE` etc.

So, I'd agree to proceed with this fix, but please add a comment there to reveal why it's necessary.

Comment (jeeminso) [reply server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: Thank you for your input. I played around with `= ANY` but couldn't find anything. `LIKE` seems to not support numeric types(uses text variant) I will follow up with this after.

Comment (jeeminso) [reply server/src/test/java/io/crate/lucene/NumericEqQueryTest.java:162]: Also GenericFunctionQuery uses NumericType.compare and it correctly handles BigDecimals with different scales.

**Review by matriv — APPROVED** [review]
thank you!

**Comment by jeeminso** [comment]
Tried out a few more approaches such as 1) prevent casting from bounded numeric to unbounded numeric for references. 2) normalize the cast by uncasting. 3) swap the cast then cast and revert the cast on the literal to identify precision loss, etc. They all turned into kind of hacky fixes.

**Pull request merged**